src/rng/src/vhu_rng.rs

```diff
@@ -0,0 +1,315 @@
+    ) -> std::result::Result<Self, std::io::Error> {
+        Ok(VuRngBackend {
+            mem: None,
+            event_idx: false,
+            random_file,
+            timer: VuRngTimerConfig::new(period_ms, max_bytes),
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| VuRngError::EventFdError)?,
+        })
+    }
+
+    pub fn process_queue(&mut self, vring: &VringRwLock) -> Result<bool> {
+        // Copy the DescriptorChains to a vector.  Processing DescriptorChains
+        // one by one directly from the Vring would lead to a deadlock because
+        // the Vring's RwLock is taken once to access the DescriptorChains and
+        // once more when vring.add_used() is called.
+        let mut desc_chains: Vec<_> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter()
+            .map_err(|_| VuRngError::DescriptorNotFound)?
+            .collect();
+
+        for desc_chain in desc_chains.iter_mut() {
+            let descriptor = match desc_chain.next() {
+                Some(descriptor) => descriptor,
+                None => continue,
+            };
+
+            let mut to_read = descriptor.len() as usize;
+            let mut timer = &mut self.timer;
+            let mut random_file = self
+                .random_file
+                .lock()
+                .map_err(|_| VuRngError::UnexpectedRngSourceAccessError)?;
+
+            if !descriptor.is_write_only() {
+                return Err(VuRngError::UnexpectedReadDescriptor);
+            }
+
+            // Get the current time
+            let now = Instant::now();
+
+            // Check how much time has passed since we started the last period.
+            match now.checked_duration_since(timer.period_start) {
+                Some(duration) => {
+                    let elapsed = duration.as_millis();
+
+                    if elapsed >= timer.period_ms {
+                        // More time has passed than a full period, reset time
+                        // and quota.
+                        timer.period_start = now;
+                        timer.quota_remaining = timer.max_bytes;
+                    } else {
+                        // If we are out of bytes for the current period.  Block until
+                        // the start of the next period.
+                        if timer.quota_remaining == 0 {
+                            let to_sleep = timer.period_ms - elapsed;
+
+                            sleep(Duration::from_millis(to_sleep as u64));
+                            timer.period_start = Instant::now();
+                            timer.quota_remaining = timer.max_bytes;
+                        }
+                    }
+                }
+                None => return Err(VuRngError::UnexpectedTimerValue),
+            };
+
+            let mem = match &self.mem {
+                Some(m) => m.memory(),
+                None => return Err(VuRngError::NoMemoryConfigured),
+            };
```

Comment:
These kind of `match`es can be simplified by using `ok_or().map_err(...)?` instead of writing the full match:
```rust
let mem = &self.mem.ok_or(VuRngError::NoMemoryConfigured)?;
```